File: src/Curves/LineCurve.ts
import { vec2 } from "gl-matrix";
import Curve from "./Curve";

export class LineCurve extends Curve {
    p0 : vec2;
    p1 : vec2;

    constructor(p0 : vec2, p1 : vec2) {
        super();
        this.p0 = p0;
        this.p1 = p1;
    }

    length() {
        let d = vec2.create();
        vec2.sub(d, this.p0, this.p1);
        return vec2.len(d);
    }

    point(t : number) {
        let result = vec2.create();
        vec2.lerp(result, this.p0, this.p1, t);
        return result;
    }

    reverse() {
        return new LineCurve(this.p1, this.p0);
    }

    split(t : number) {
        let m = this.point(t);
        let res : [LineCurve, LineCurve] = [new LineCurve(this.p0, m), new LineCurve(m, this.p1)];
        return res;
    }
}

File: Assets/Scripts/Geom/LineCurve.cs
using UnityEngine;

namespace Geom
{
    public class LineCurve : ICurve
    {
        public Vector2 P0;
        public Vector2 P1;

        public LineCurve(Vector2 p0, Vector2 p1)
        {
            P0 = p0;
            P1 = p1;
        }

        public Vector2 AsVector()
        {
            return P1 - P0;
        }

        public float Length()
        {
            return AsVector().magnitude;
        }

        public Vector2 Point(float t)
        {
            return Vector2.Lerp(P0, P1, t);
        }

        public (ICurve, ICurve) Split(float t)
        {
            Vector2 point = Point(t);
            return (new LineCurve(P0, point), new LineCurve(point, P1));
        }
    }
}